lib/segment/src/segment.rs

```diff
@@ -209,16 +213,46 @@ impl Segment {
 
     /// Retrieve payload by internal ID
     #[inline]
-    fn payload_by_offset(
-        &self,
-        point_offset: PointOffsetType,
-    ) -> OperationResult<TheMap<PayloadKeyType, PayloadType>> {
+    fn payload_by_offset(&self, point_offset: PointOffsetType) -> OperationResult<Payload> {
         Ok(self.payload_storage.borrow().payload(point_offset))
     }
 
     pub fn save_current_state(&self) -> OperationResult<()> {
         self.save_state(&self.get_state())
     }
+
+    fn infer_from_payload_data(&self, key: PayloadKeyTypeRef) -> Option<PayloadSchemaType> {
+        let payload_store = self.payload_storage.borrow();
+        match payload_store.iter_ids().next() {
+            Some(id) => {
+                let payload = payload_store.payload(id);
+                Some(payload.get_schema_type(key))
+            }
+            None => None,
+        }
+    }
+
+    fn check_schema(&self, payload: &Payload) -> OperationResult<()> {
+        for (field_name, schema) in self.schema_storage.as_map() {
+            match schema {
+                Some(payload_type) => match payload_type {
+                    PayloadSchemaType::Unknown => Ok(()),
+                    _ => payload.get_checked(field_name, &payload_type)?,
+                },
+                None => {
+                    let payload_type = payload.infer_type(field_name.into());
+                    self.schema_storage
+                        .update_schema_value(&field_name, payload_type)?;
+                    if payload_type != PayloadSchemaType::Unknown {
```

Comment:
What is the difference between `payload_type == None` and `payload_type == PayloadSchemaType::Unknown` ?